### ISSUE TITLE
Running QEMU process in a confined environment

### ISSUE THREAD

**Issue by fengwang666** [kata-containers/kata-containers#2321]
**Is your feature request related to a problem? Please describe.**

We're evaluating to use Kata container with QEMU in our production environment but are a bit concerned with the security risk of QEMU. QEMU is known to have a large attack surface and so far [24 CVEs](https://www.cvedetails.com/product/12657/Qemu-Qemu.html?vendor_id=7506) have been reported this year.  

**Describe the solution you'd like**

To mitigate the security risk, one option we are thinking is to run the QEMU process in a confined environment using Linux primitives, such as namespace, seccomp, and cgroup, similar to what [firecracker](https://github.com/firecracker-microvm/firecracker/blob/main/docs/jailer.md) does. I did some initial prototyping with the Firecracker jailer and it looks promising. I'd like to gather feedback and suggestions from the community on this approach. Has anyone tried something similar on QEMU and are there any caveats or blockers? 

If we're okay with the approach, I'm happy to implement it. Another question regarding the process, besides following the [contributing guideline](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md) and submitting the PR, anything else I should be aware?

**Describe alternatives you've considered**

Firecracker and Cloud Hypervisor seem good alternatives from a security perspective. The challenges are Firecracker lacks some features we need and cloud hypervisor is relatively new and it's not clear whether it's ready for production use. 

**Additional context**

Found a related issue: https://github.com/kata-containers/runtime/issues/345

**Before raising this feature request**

Have you looked at the [limitations document](https://github.com/kata-containers/documentation/blob/master/Limitations.md)? Yes

**Comment by bergwolf** [comment]
@fengwang666 Yes, jailer is a good starting point to make qemu more secure.

Please note that jailer doesn't support passing command line arguments (except for a ``--config-file` option) to the running program, but qemu usually need many. So we'd have to either patch jailer to support it, or we can create a shell file with all qemu commandlines and pass it to jailer.

Also we need to create a kata user to run jailer as non-root. We can make it an option in the global config file, defaulting to root as it requires sysadmin to grant proper privilege to the user to access /dev/kvm and /dev/vhost-net/vsock etc.

When implementing it, please make sure it is hypervisor agnostic so that cloud hypervisor can use it as well.

**Comment by fengwang666** [comment]
@bergwolf Thanks for the prompt reply. Can you clarify the statement `run jailer as non-root`? Are you suggesting to create a privilege user to run jailer instead of using root? AFAIK, the jailer process needs to have some privileged capabilities to call chroot, create device node, and mount files.

**Comment by c3d** [comment]
In distros such as Fedora, libvirt is typically used to "jail" qemu and run it as a non-root user. This is quite complicated, notably regarding safe access of the devices that qemu needs. Some of the advanced features of qemu are difficult to jail correctly. If you are interested, I can try to find the work that was done attempting to run qemu through libvirt back in 2020.

We also build a custom configuration of qemu that removes many devices and reduces the attack surface quite a bit. I believe that out of the 24 CVEs you quoted, only 2 are applicable with the configuration we have (the two virtiofs ones), and one was applicable to kata 1.x (the 9pfs issue).

**Comment by fengwang666** [comment]
First, thanks for all the great feedback.

@c3d , I'd love to hear more about the libvirt work you mentioned, especially on the advanced qemu features difficult to jail.  I also noticed a lot of the qemu libraries are excluded in the custom qemu build and I appreciate that.

@devimc , for your second point, is the issue that a newly created mount point is not populated into the QEMU mount namespace? I can see this become a problem for hot-plug devices. But can you further explain why this will block the `PODs with several containers use case`?

**Comment by devimc** [comment]
> for your second point, is the issue that a newly created mount point is not populated into the QEMU mount namespace? 

right, mnt namespace is like snapshot of the current list of devices and mount points available in the host when it is created

> I can see this become a problem for hot-plug devices. But can you further explain why this will block the PODs with several containers use case?

the mnt namespace is created when the first container is created (pause container) and QEMU will be placed there, this mnt namespace won't have access (won't see) to other container's mnt namespaces

**Comment by jodh-intel** [comment]
Related: https://github.com/kata-containers/kata-containers/blob/main/docs/hypervisors.md.

**Comment by amshinde** [comment]
@fengwang666 You can constrain qemu using selinux policies, there is support in Kata to do so. On enabling selinux, selinux policies are applied to the kvm launcher tool.
In addition there is a PR in flight for enabling seccomp support for qemu. PTAL: https://github.com/kata-containers/kata-containers/pull/2312

This is a good guide outlining the isolation mechanisms that can be applied to qemu. In Kata, the qemu process is run as root, but if you dont need to support device passthrough, you can get away with running qemu as non-root user and configuring  UNIX groups for access to /dev/kvm, /dev/net/tun, and other device nodes as described in that doc.

**Comment by fengwang666** [comment]
Thanks. @amshinde 
Can you provide a link to the guide you mentioned?

**Comment by bergwolf** [comment]
@fengwang666 sorry for not being clear. I meant to say to `run qemu as non-root with jailer` ;)

Also we did discuss this during the AC meeting today. The libvirt work is currently on hold from Red Hat side. See the meeting notes at https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs

**Comment by fengwang666** [comment]
@bergwolf @c3d @devimc @amshinde Has there been any discussion on running qemu inside a runc container as a security confinement? Is it because runc has a large attack surface?

**Comment by devimc** [comment]
@fengwang666 
> Wouldn't it be easier to get cgroup, namespace and other security isolations from runc than a custom implementation?

you'll end up facing the same issue I mentioned above, the hypervisor should be part of all containers' namespaces and afaik that's not possible

**Comment by fengwang666** [comment]
@bergwolf @c3d @devimc @amshinde I create a [draft PR](https://github.com/kata-containers/kata-containers/pull/2432) to show how to enable running QEMU as a non-root process. I appreciate your feedback on the approach.
